### PR TITLE
Systemd socket activation

### PR DESCRIPTION
**Pull request by xaiki**
This tiny patch implements systemd's socket activation mechanism[0] for
cherrypy servers, it's based on work sponsored by Endless Computers[1] to
package kalite[2] as a flatpak[3].

Socket Activation allows to setup a system so that systemd will sit on a
port and start services 'on demand' (a little bit like inetd and xinetd
used to do). This patch makes that mechanism work for any cherrypy
service without modification.

[0] http://0pointer.de/blog/projects/socket-activation.html
[1] https://endlessm.com
[2] https://learningequality.org/ka-lite/
[3] https://flatpak.org

https://phabricator.endlessm.com/T4489

Signed-off-by: Niv Sardi xaiki@evilgiggle.com

### CODE REVIEW

File: cherrypy/wsgiserver/__init__.py [hunk 1]
```diff
@@ -1939,7 +1939,11 @@ def start(self):
             self.software = "%s Server" % self.version
 
         # Select the appropriate socket
-        if isinstance(self.bind_addr, six.string_types):
+        self.socket = None
+        if os.environ.get('LISTEN_PID', None):
```
Comment (webknjaz) [inline cherrypy/wsgiserver/__init__.py]: Why set default value to `None`? It's not necessary.
Moreover, [`os.getenv('LISTEN_PID')`](https://docs.python.org/2/library/os.html#os.getenv) would be more verbose.

File: cherrypy/wsgiserver/__init__.py [hunk 2]
```diff
@@ -1973,21 +1977,21 @@ def start(self):
...
+                af, socktype, proto, canonname, sa = res
+                try:
+                    self.bind(af, socktype, proto)
+                except socket.error, serr:
+                    msg = "%s -- (%s: %s)" % (msg, sa, serr)
+                    if self.socket:
+                        self.socket.close()
+                    self.socket = None
+                    continue
+                break
```
Comment (webknjaz) [inline cherrypy/wsgiserver/__init__.py]: I'd move `break` to line 1986, just next to self.bind() and remove `continue`.

File: cherrypy/wsgiserver/__init__.py [hunk 3]
```diff
@@ -1973,21 +1977,21 @@ def start(self):
...
-                continue
-            break
         if not self.socket:
-            raise socket.error(msg)
+            msg = "No socket could be created"
+            for res in info:
+                af, socktype, proto, canonname, sa = res
+                try:
+                    self.bind(af, socktype, proto)
+                except socket.error, serr:
```
Comment (webknjaz) [inline cherrypy/wsgiserver/__init__.py]: This outdated exception handling syntax breaks TravisCI tests. Plz fix

File: cherrypy/wsgiserver/__init__.py (hunk 1 above)
Comment (xaiki) [inline cherrypy/wsgiserver/__init__.py]: fixed

File: cherrypy/wsgiserver/__init__.py (hunk 2 above)
Comment (xaiki) [inline cherrypy/wsgiserver/__init__.py]: fixed

File: cherrypy/wsgiserver/__init__.py (hunk 3 above)
Comment (xaiki) [inline cherrypy/wsgiserver/__init__.py]: fixed

**Comment by xaiki** [comment]
@webknjaz thanks for the review, i fixed your comments and pushed !

**Comment by webknjaz** [comment]
LGTM, but there's some issue w/ [TravisCI](https://travis-ci.org/cherrypy/cherrypy/builds/146641684). Could you plz try restarting builds?
I think we can now ping @jaraco and see if he would merge this.

**Comment by jaraco** [comment]
I tried restarting the builds, but the result is the same. Why would the tests not run on Travis with this patch?

**Comment by webknjaz** [comment]
@jaraco I think, there's some issue with TravisCI itself. Look at the `pip install` output. It differs for different jobs. Did you mention this?
I have a strong feeling that it gets cut/halted at some point.

**Comment by webknjaz** [comment]
@jaraco I've found the same issue, which happened in the past: https://travis-ci.org/cherrypy/cherrypy/jobs/139925725

**Comment by webknjaz** [comment]
Okay.. I've tried running `nosetests -s -v` myself and I also get return code 70. So the issue is not in Travis

**Comment by webknjaz** [comment]
``` sh
$ .tox/py27/bin/nosetests -s -v
nose.config: INFO: Set working dir to ~/src/cherrypy-xaiki/cherrypy
nose.config: INFO: Working directory ~/src/cherrypy-xaiki/cherrypy is a package; adding to sys.path
nose.config: INFO: Ignoring files matching ['^\\.', '^_', '^setup\\.py$']
nose.selector: INFO: ~/src/cherrypy-xaiki/cherrypy/cherryd is executable; skipped
nose.selector: INFO: ~/src/cherrypy-xaiki/cherrypy/daemon.py is executable; skipped
nose.selector: INFO: ~/src/cherrypy-xaiki/cherrypy/test/sessiondemo.py is executable; skipped
nose.selector: INFO: ~/src/cherrypy-xaiki/cherrypy/test/test_session.py is executable; skipped
$ echo $?
70
```

**Comment by webknjaz** [comment]
It looks there's something malformed in this patch. Even if I generate diff and apply it to working master it breaks nose command.

**Comment by xaiki** [comment]
@webknjaz it was the removal of `self.socket = None` i added it back and it now unbreaks nose, i have no idea why ....

**Comment by webknjaz** [comment]
@xaiki okay, seems good

Could you plz include some documentation/examples of running cherrypy app with use of this feature?

Also it would be nice to have some tests covering it.

**Comment by xaiki** [comment]
a test is a bit hard as you need to have systemd setup, 

we (endlessm.com) use that patch to enable kalite systemd activation, here's our system part:
https://github.com/mariospr/eos-kalite-system-helper

basically, you need to setup these files:
https://github.com/mariospr/eos-kalite-system-helper/blob/master/debian/eos-kalite-system-helper.service
https://github.com/mariospr/eos-kalite-system-helper/blob/master/debian/eos-kalite-system-helper.socket

to make it work

**Comment by webknjaz** [comment]
@xaiki Could you plz describe this in [docs/deploy.rst](/cherrypy/cherrypy/blob/master/docs/deploy.rst)?
Ability of others to use this feature seamlessly will make this PR more valuable.
Also add some information about the feature to the docstring of `start()` method.

**Comment by xaiki** [comment]
docs added to deploy.rst
for start you mean start of the wcgiserver ? or the dameon start ? i'm not that familiar with those uses =)

**Comment by xaiki** [comment]
added to the bind_addr docstring as it seemed more relevant.

**Comment by webknjaz** [comment]
LGTM

@jaraco I thinks this PR is ready for merge

**Comment by xaiki** [comment]
good, i had missed a patch from @mariospr just got it in the PR. thanks for a great review @webknjaz

**Comment by jaraco** [comment]
Excellent. Thanks everyone for putting this together.

**Comment by jaraco** [comment]
Pending release in 7.1.0.

**Comment by mariospr** [comment]
@xaiki, @jaraco: that was fast! Thanks for accepting my patch :-)

**Pull request merged**